File: app/src/main/java/com/plbear/iweight/model/details/DetailsAdapter.java
package com.plbear.iweight.model.details;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckBox;
import android.widget.TextView;

import com.plbear.iweight.R;
import com.plbear.iweight.data.Data;
import com.plbear.iweight.data.DataManager;
import com.plbear.iweight.utils.LogInfo;
import com.plbear.iweight.utils.Utils;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * Created by yanyongjun on 2018/6/28.
 */

public class DetailsAdapter extends BaseAdapter {
    private final static String TAG = "DetailsAdapter";
    private Context mContext;
    private ArrayList<Data> mListData;
    private LayoutInflater mInflater;
    boolean isEditMode = false;
    private HashMap<Integer, Boolean> mSelectMap = new HashMap<>();
    private int mSelectCount = 0;
    private DetailsAdapter.OnItemClick mCallback;
    private ArrayList<Data> selectData;

    public ArrayList<Data> getSelectData() {
        ArrayList<Data> list = new ArrayList<>(mSelectCount);
        java.util.Iterator<Integer> it = mSelectMap.keySet().iterator();
        while (it.hasNext()) {
            int key = it.next();
            if (mSelectMap.get(key)) {
                list.add(mListData.get(key));
            }
        }
        return list;
    }

    public interface OnItemClick {
        void itemClick(HashMap<Integer, Boolean> mSelectMap, int selectCount);

        void longItemClick(boolean editMode);
    }

    public void notifyDataSetChanged() {
        DataManager db = DataManager.getInstance(mContext);
        mListData = db.queryAll();
        Utils.sortDataBigToSmall(mListData);
        super.notifyDataSetChanged();
    }


    public DetailsAdapter(Context context, ArrayList<Data> list, DetailsAdapter.OnItemClick callback) {
        mContext = context;
        mListData = list;
        mCallback = callback;
        mInflater = LayoutInflater.from(mContext);
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        View view;
        if (isEditMode) {
            view = mInflater.inflate(R.layout.item_details_editmode, null);
        } else {
            view = mInflater.inflate(R.layout.item_details, null);
        }

        TextView labDate = view.findViewById(R.id.lab_details_item_date);
        TextView labWeight = view.findViewById(R.id.lab_details_item_weight);
        Data data = mListData.get(position);
        labDate.setText(Utils.formatTimeFull(data.getTime()));
        labWeight.setText("" + data.getWeight());
        if (isEditMode) {
            final CheckBox selectBox = view.findViewById(R.id.checkbox_details_item_select);
            selectBox.setVisibility(View.VISIBLE);
            selectBox.setOnClickListener(null);
            selectBox.setClickable(false);
            Boolean isCheck = mSelectMap.get(position);
            selectBox.setClickable(isCheck != null && true == isCheck);
            selectBox.setChecked(isCheck != null && true == isCheck);
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Boolean temp = mSelectMap.get(position);
                    boolean click = temp != null && true == temp;
                    mSelectMap.put(position, !click);
                    if (click) {
                        mSelectCount--;
                    } else {
                        mSelectCount++;
                    }
                    selectBox.setChecked(!click);
                    mCallback.itemClick(mSelectMap, mSelectCount);
                }
            });
        } else {
            view.setClickable(false);
            view.setOnClickListener(null);
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    enterEditMode();
                }
            });
        }
        return view;
    }

    public long getItemId(int position) {
        return (long) position;
    }

    public Object getItem(int position) {
        return mListData.get(position);
    }

    public int getCount() {
        return mListData.size();
    }


    void enterEditMode() {
        if (!isEditMode) {
            isEditMode = true;
            mSelectMap.clear();
            notifyDataSetChanged();
        }
        mCallback.longItemClick(isEditMode);
    }

    void exitEditMode() {
        if (isEditMode) {
            isEditMode = false;
            mSelectMap.clear();
            mSelectCount = 0;
            notifyDataSetChanged();
        }
        mCallback.longItemClick(isEditMode);
    }

    /**
     * on select all click
     */
    void onSelectAllClick() {
        boolean isSelectAll = mSelectCount == mListData.size();
        if (isSelectAll) {
            mSelectMap.clear();
            mSelectCount = 0;
        } else {
            int i = 0;
            for (Data data : mListData) {
                mSelectMap.put(i++, true);
            }
            mSelectCount = mListData.size();
        }
        notifyDataSetChanged();
        mCallback.itemClick(mSelectMap, mSelectCount);
    }

}

File: app/src/main/java/com/plbear/iweight/model/details/DetailsAdapter.kt
package com.plbear.iweight.model.details

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.CheckBox
import android.widget.TextView

import com.plbear.iweight.R
import com.plbear.iweight.data.Data
import com.plbear.iweight.data.DataManager
import com.plbear.iweight.utils.MyLog
import com.plbear.iweight.utils.Utils

import java.util.ArrayList
import java.util.HashMap

/**
* Created by yanyongjun on 2016/11/23.
*/

class DetailsAdapter(context: Context, list: ArrayList<Data>, callback: OnItemClick) : BaseAdapter() {
    private var mContext = context
    private var mListData = list
    private var mInflater: LayoutInflater = LayoutInflater.from(context)
    var isEditMode = false
    private val mSelectMap = HashMap<Int, Boolean>()
    private var mSelectCount = 0
    private var mCallback: OnItemClick = callback

    val selectData: ArrayList<Data>
        get() {
            val list = ArrayList<Data>(mSelectCount)
            val it = mSelectMap.keys.iterator()
            while (it.hasNext()) {
                val key = it.next()
                if (mSelectMap[key] == true) {
                    list.add(mListData[key])
                }
            }
            return list
        }

    interface OnItemClick {
        fun itemClick(mSelectMap: HashMap<Int, Boolean>, selectCount: Int)

        fun longItemClick(editMode: Boolean)
    }

    override fun notifyDataSetChanged() {
        val db = DataManager.getInstance(mContext)
        mListData = db!!.queryAll()
        Utils.sortDataBigToSmall(mListData)
        super.notifyDataSetChanged()
    }


    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        var view: View
        if (isEditMode) {
            view = mInflater.inflate(R.layout.item_details_editmode, null)
        } else {
            view = mInflater.inflate(R.layout.item_details, null)
        }
        val labDate = view.findViewById<TextView>(R.id.lab_details_item_date)
        val labWeight = view.findViewById<TextView>(R.id.lab_details_item_weight)
        val data = mListData[position]
        labDate.text = Utils.formatTimeFull(data.time)
        labWeight.setText(""+data.weight)
        if (isEditMode) {
            val selectBox = view.findViewById<View>(R.id.checkbox_details_item_select) as CheckBox
            selectBox.visibility = View.VISIBLE
            selectBox.setOnClickListener(null)
            selectBox.isClickable = false
            val isCheck = mSelectMap[position]
            selectBox.isChecked = isCheck == true
            view.setOnClickListener {
                val temp = mSelectMap[position]
                val click = temp ?: false
                mSelectMap.put(position, !click)
                if (click) {
                    mSelectCount--
                } else {
                    mSelectCount++
                }
                selectBox.isChecked = !click
                mCallback.itemClick(mSelectMap, mSelectCount)
            }
        } else {
            view.isClickable = false
            view.setOnClickListener(null)
            view.setOnLongClickListener {
                enterEditMode()
                true
            }
        }
        return view
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getItem(position: Int): Any? {
        return mListData[position]
    }

    override fun getCount(): Int {
        MyLog.e(TAG,"yanlog getcount:"+mListData.size)
        return mListData.size
    }

    fun enterEditMode() {
        if (!isEditMode) {
            isEditMode = true
            mSelectMap.clear()
            notifyDataSetChanged()
        }
        mCallback.longItemClick(isEditMode)
    }

    fun exitEditMode() {
        if (isEditMode) {
            isEditMode = false
            mSelectMap.clear()
            mSelectCount = 0
            notifyDataSetChanged()
        }
        mCallback.longItemClick(isEditMode)

    }

    /**
     * on select all click
     */
    fun onSelectAllClick() {
        val isSelectAll = mSelectCount == mListData!!.size
        if (isSelectAll) {
            mSelectMap.clear()
            mSelectCount = 0
        } else {
            for (i in mListData!!.indices) {
                mSelectMap.put(i, true)
            }
            mSelectCount = mListData!!.size
        }
        notifyDataSetChanged()
        mCallback.itemClick(mSelectMap, mSelectCount)
    }

    companion object {
        private val TAG = "DetailsAdapter"
    }
}
